### ISSUE TITLE
Issue about the rasterizer and nu degree

### ISSUE THREAD

**Issue by delaynomore-20** [realcrane/3D-student-splatting-and-scooping#7]
Dear authors,
First, thank you for sharing this excellent repository—it’s been really helpful for my rendering work! I’ve been modifying your codebase to adapt it for 2D rendering tasks, but I’ve run into a problem with the nu_degree parameter and would appreciate your guidance.
### **Problem Description**
During training, I noticed two key issues with the nu_degree parameter:
No parameter updates: The value of nu_degree stays the same throughout the training process (it remains at its initial value, never changing).
Zero gradient: When I check the gradient of nu_degree after loss.backward(), it’s consistently None or 0 (I’ve confirmed nu_degree.requires_grad is set to True).
### **My Troubleshooting Steps**
To diagnose this, I:
Verified nu_degree.requires_grad is True (so gradient tracking is enabled).
Checked the training loop: Loss is computed normally, and other parameters (e.g., _opacity) update as expected with valid gradients.
Looked into the t_render file (the rendering module) and noticed that nu_degree doesn’t seem to be passed into the renderer or used in the rendering pipeline.
`    def forward(self, means3D, means2D, opacities, nu_degree, cover, negative_value, shs = None, colors_precomp = None, scales = None, rotations = None, cov3D_precomp = None):
        
        raster_settings = self.raster_settings

        if (shs is None and colors_precomp is None) or (shs is not None and colors_precomp is not None):
            raise Exception('Please provide excatly one of either SHs or precomputed colors!')
        
        if ((scales is None or rotations is None) and cov3D_precomp is None) or ((scales is not None or rotations is not None) and cov3D_precomp is not None):
            raise Exception('Please provide exactly one of either scale/rotation pair or precomputed 3D covariance!')
        
        if shs is None:
            shs = torch.Tensor([])
        if colors_precomp is None:
            colors_precomp = torch.Tensor([])

        if scales is None:
            scales = torch.Tensor([])
        if rotations is None:
            rotations = torch.Tensor([])
        if cov3D_precomp is None:
            cov3D_precomp = torch.Tensor([])

        # Invoke C++/CUDA rasterization routine
        return rasterize_t(
            means3D,
            means2D,
            shs,
            colors_precomp,
            opacities,
            scales, 
            rotations,
            cov3D_precomp,
            raster_settings,
            nu_degree,
            cover,
            negative_value
        )`
### **My Guess (and Possible Misunderstanding)**
I suspect the root cause is that nu_degree isn’t integrated into the rendering process. Since the parameter doesn’t contribute to generating the 2D rendered output (and thus doesn’t affect the loss calculation), its gradient remains zero and it never updates.
However, this could also be due to my limited understanding of your code’s core logic—maybe nu_degree is supposed to be used in a part of the pipeline I missed, or there’s a specific way to include it for 2D rendering.
### **Request for Guidance**
Could you help clarify:
Whether nu_degree needs to be explicitly passed into the renderer (from t_render) to participate in 2D rendering?
If there are other parts of the codebase I should modify to ensure nu_degree contributes to the loss calculation (and thus gets updated)?
Let me know if you need more details about my modifications or reproducible steps—I’m happy to share!
Thanks again for your work and support.

**Comment by delaynomore-20** [comment]
To clarify, my changes to the codebase have been minimal: I only added necessary code in the NTModel class and dataset-related files to support my 2D rendering task. All other parts of the original codebase (including the rendering logic in t_render, training loops, and other core modules) remain unmodified.
